sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala

```diff
@@ -0,0 +1,451 @@
+import scala.util.Try
+
+import org.apache.hadoop.conf.Configuration
+import org.rocksdb.{RocksDB => NativeRocksDB, _}
+
+import org.apache.spark.TaskContext
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.catalyst.util.CaseInsensitiveMap
+import org.apache.spark.util.{NextIterator, Utils}
+
+/**
+ * Class representing a RocksDB instance that checkpoints version of data to DFS.
+ * After a set of updates, a new version can be committed by calling `commit()`.
+ * Any past version can be loaded by calling `load(version)`.
+ *
+ * @note This class is not thread-safe, so use it only from one thread.
+ * @see [[RocksDBFileManager]] to see how the files are laid out in local disk and DFS.
+ * @param dfsRootDir  Remote directory where checkpoints are going to be written
+ * @param conf         Configuration for RocksDB
+ * @param localRootDir Root directory in local disk that is used to working and checkpointing dirs
+ * @param hadoopConf   Hadoop configuration for talking to the remote file system
+ * @param loggingId    Id that will be prepended in logs for isolating concurrent RocksDBs
+ */
+class RocksDB(
+    dfsRootDir: String,
+    val conf: RocksDBConf,
+    localRootDir: File = Utils.createTempDir(),
+    hadoopConf: Configuration = new Configuration,
+    loggingId: String = "") extends Logging {
+
+  RocksDBLoader.loadLibrary()
+
+  // Java wrapper objects linking to native RocksDB objects
+  private val readOptions = new ReadOptions()  // used for gets
+  private val writeOptions = new WriteOptions().setSync(true)  // wait for batched write to complete
+  private val flushOptions = new FlushOptions().setWaitForFlush(true)  // wait for flush to complete
+  private val writeBatch = new WriteBatchWithIndex(true)  // overwrite multiple updates to a key
+
+  private val bloomFilter = new BloomFilter()
+  private val tableFormatConfig = new BlockBasedTableConfig()
+  tableFormatConfig.setBlockSize(conf.blockSizeKB * 1024)
+  tableFormatConfig.setBlockCache(new LRUCache(conf.blockCacheSizeMB * 1024 * 1024))
+  tableFormatConfig.setFilterPolicy(bloomFilter)
+
+  private val dbOptions = new Options() // options to open the RocksDB
+  dbOptions.setCreateIfMissing(true)
+  dbOptions.setTableFormatConfig(tableFormatConfig)
+  private val dbLogger = createLogger() // for forwarding RocksDB native logs to log4j
+  dbOptions.setStatistics(new Statistics())
+
+  private val workingDir = createTempDir("workingDir")
+  private val fileManager = new RocksDBFileManager(
+    dfsRootDir, createTempDir("fileManager"), hadoopConf, loggingId = loggingId)
+  private val byteArrayPair = new ByteArrayPair()
+  private val commitLatencyMs = new mutable.HashMap[String, Long]()
+  private val acquireLock = new Object
+
+  @volatile private var db: NativeRocksDB = _
+  @volatile private var loadedVersion = -1L   // -1 = nothing valid is loaded
+  // number of keys in all committed versions before
+  @volatile private var numCommittedKeys = 0L
```

Comment:
why it is for "all committed versions"? In `load`, we just overwrite it `numCommittedKeys = metadata.numKeys`. So looks like it is only for the loaded version.

Reply:
It's also used in the `commit` operation, which will be introduced in the next commit: https://github.com/apache/spark/pull/32933/files#diff-4663b8349511dc040c3d5204bb6868064002d0b45aadaf1f1795a683897a5c7dR224

Reply:
Here "all" seems to make confusion. The number only represents for "one" version, hence the confusion.

Now I think it'd be nice if we can rename the field name to represent these fields are representing the number of keys per certain version. I guess below candidates could help avoiding confusion:

* numKeysOnLoadedVersion - numKeysOnWritingVersion
* numKeysOnCommittedVersion - numKeysOnUncommittedCommittingVersion
* or better names

Reply:
+1 for the `numKeysOnLoadedVersion - numKeysOnWritingVersion`. Reviewed the new name in all the places (includes the further PRs) and it has a clear meaning. Thanks for the advice!

Reply:
looks good to me.